oca/api.py

```diff
@@ -49,15 +32,66 @@ def method(self):
 
             [1, 2, 3]
 
+        If no container is provided as an argument to the decorator, and a
+        value is returned from the decorated method, a ValueError will be
+        raised.
+
+        .. code-block:: python
+
+            @api.foreach
+            def method(self):
+                # Raises ValueError
+                return 1+2
+
+            @api.foreach
+            def method(self):
+                # No error raised
+                self.math = 1+2
+
+    Args:
+        container (type): Type of container to wrap results in. It will be
+            instantiated with an iterable of the decorated method results.
+
+    Raises:
+        ValueError: If no container type is provided, and a value is returned
```

Comment:
If you follow my suggestion below, this would be `AssertionError`, and you should advise it would only happen with `PYTHONOPTIMIZE<2`. But please continue reading. I'd also add a hint about using `@api.multi` if you really need to return something, because it can mean some better performance improvements in many cases.